internal/storage/leveldb.go

```diff
@@ -72,46 +127,80 @@ func (s *LevelDBStorage) GetLastQueryHeight(queryID uint64) (block uint64, found
+//  2. tx submitted successfully (temporary status, should be updated after neutron tx committed into the block) - relay.Submitted
+//     2.a) failed to commit tx into the block - relay.ErrorOnCommit
+//     2.b) tx successfully committed - relay.Committed
+//
 // To convert status from "2" to either "2.a" or "2.b" we use additional SubmittedTxStatusPrefix storage to track txs
-func (s *LevelDBStorage) SetTxStatus(queryID uint64, hash string, neutronHash string, status relay.SubmittedTxInfo) (err error) {
-	s.Lock()
-	defer s.Unlock()
+func (s *LevelDBStorage) SetTxStatus(queryID uint64, hash string, neutronHash string, txInfo relay.SubmittedTxInfo, processedTx *relay.Transaction) (err error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 
-	// save tx status
 	t, err := s.db.OpenTransaction()
 	if err != nil {
 		return fmt.Errorf("failed to open leveldb transaction: %w", err)
 	}
 
 	defer t.Discard()
-	data, err := json.Marshal(status)
+	data, err := json.Marshal(txInfo)
 	if err != nil {
 		return fmt.Errorf("failed to Marshal SubmittedTxInfo: %w", err)
 	}
 
-	err = t.Put(constructKey(queryID, hash), data, nil)
+	err = t.Put(constructTxStatusKey(queryID, hash), data, nil)
 	if err != nil {
-		return fmt.Errorf("failed to set tx status: %w", err)
+		return fmt.Errorf("failed to set tx txInfo: %w", err)
+	}
+
+	if processedTx != nil {
+		err = cacheProcessedTx(t, queryID, hash, processedTx)
+		if err != nil {
+			return fmt.Errorf("failed to cache processed tx: %w", err)
+		}
 	}
 
-	if status.Status == relay.Submitted {
-		txInfo := relay.PendingSubmittedTxInfo{
+	if txInfo.Status == relay.Submitted {
+		pendingTxInfo := relay.PendingSubmittedTxInfo{
 			QueryID:         queryID,
 			SubmittedTxHash: hash,
 			NeutronHash:     neutronHash,
 		}
-		err = saveIntoPendingQueue(t, neutronHash, txInfo)
+		err = saveIntoPendingQueue(t, neutronHash, pendingTxInfo)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to save txInfo into pending queue: %w", err)
 		}
-	} else if status.Status == relay.Committed || status.Status == relay.ErrorOnCommit {
+	} else if txInfo.Status == relay.Committed || txInfo.Status == relay.ErrorOnCommit {
 		err = removeFromPendingQueue(t, neutronHash)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to remove txInfo from pending queue: %w", err)
+		}
+	}
+
+	if txInfo.Status == relay.ErrorOnCommit || txInfo.Status == relay.ErrorOnSubmit {
+		unsuccessfulTxInfo := relay.UnsuccessfulTxInfo{
+			QueryID:         queryID,
+			SubmittedTxHash: hash,
+			NeutronHash:     neutronHash,
+			ErrorTime:       time.Now(),
+			Status:          txInfo.Status,
+			Message:         txInfo.Message,
+		}
+		err = saveIntoUnsuccessfulQueue(t, queryID, hash, unsuccessfulTxInfo)
+		if err != nil {
+			return fmt.Errorf("failed to save unsuccessfulTxInfo into Unsuccessful queue: %w", err)
+		}
+	}
+
+	if txInfo.Status == relay.Committed {
+		err = removeFromUnsuccessfulQueue(t, queryID, hash)
```

Comment:
`removeFromUnsuccessfulQueue` doesn't return an error if tx is not in the UnsuccessfulQueue, right?
So, there are no problems with this `removeFromUnsuccessfulQueue` call, if a tx was not saved into `UnsuccessfulQueue` and immediately goes with `relay.Committed` status here?